tests/handler/handler_copy.cpp

```diff
@@ -17,334 +22,887 @@ using mode_t = cl::sycl::access::mode;
+    dstBuf = std::unique_ptr<buffer_dst_t>(new buffer_dst_t(dstBufRange));
+
+    fill_buffer(queue, *srcBuf, encode_index_init_op<dataT, dim_src>());
+    fill_buffer(queue, *dstBuf, scalar_init_op<dataT, dim_dst>(deviceCanary));
+
+    queue.wait_and_throw();
   }
 
-/**
- * @brief Checks whether copy functions that accept a write accessor work.
- */
-#define TEST_SINGLE_COPY_FUNCTION_ACC_WRITE(dataT, dim_src, dim_dst, readMode, \
-                                            writeMode, target, queue, func)    \
-  {                                                                            \
-    test_single_copy_function_base<dataT, dim_src, dim_dst> testBase;          \
-    testBase.reset_data();                                                     \
-    queue.submit([&](cl::sycl::handler& cgh) {                                 \
-      auto accRead = accessor_helper<dataT, dim_src, readMode, target>::get(   \
-          cgh, testBase.m_bufRead);                                            \
-      auto accWrite = accessor_helper<dataT, dim_dst, writeMode, target>::get( \
-          cgh, testBase.m_bufWrite);                                           \
-      func(cgh, accRead, accWrite, testBase.m_hostPtrRead,                     \
-           testBase.m_hostPtrWrite);                                           \
-    });                                                                        \
-    queue.wait_and_throw();                                                    \
+  /**
+   * @brief Verifies a device to host copy.
+   *
+   * While the device source may be strided, the target host memory region is
+   * always dense.
+   */
+  template <typename test_fn>
+  void verify_d2h_copy(test_fn fn, const log_helper& lh) const {
+    run_test_function(fn, lh);
+
+    for (size_t i = 0; i < numElems; ++i) {
+      const auto received = dstHostPtr.get()[i];
+
+      if (i < srcCopyRange.size()) {
+        const auto srcRelIdx = reconstruct_index(srcCopyRange, i);
+        const auto expected =
+            encode_index_init_op<dataT, dim_src>{}(srcCopyOffset + srcRelIdx);
+
+        if (!th::equal(received, expected)) {
+          log_error(lh, cl::sycl::id<3>(i, 0, 0), received, expected);
+          return;
+        }
+      } else {
+        if (!th::equal(received, hostCanary)) {
+          log_canary_violation(lh, cl::sycl::id<3>(i, 0, 0), received);
+          return;
+        }
+      }
+    }
+  }
+
+  /**
+   * @brief Verifies that the host memory backing the source buffer has been
+   * updated correctly.
+   *
+   * Note that we don't check any canary values outside of the updated region
+   * (when using a ranged accessor) as a SYCL implementation is free to also
+   * update other parts of the host memory.
+   */
+  template <typename test_fn>
+  void verify_update_host(test_fn fn, const log_helper& lh) const {
+    run_test_function(fn, lh);
+
+    std::lock_guard<cl::sycl::mutex_class> lock(srcBufHostMemoryMutex);
```

Comment:
Why do you need this?

Reply:
Ah yes, I have "open issue about shared ownership host-buffer usage semantics in SYCL-Docs repo" somewhere in my backlog... It was my understanding from section 4.7.5 that the host memory region is only guaranteed to be updated once the mutex has been released by the SYCL runtime.